tools/windows/DatadogAgentInstaller/CustomActions/RollbackDataStore.cs

```diff
@@ -0,0 +1,219 @@
+
+        public RollbackDataStore(ISession session, string name, IFileSystemServices fileSystemServices)
+        {
+            _session = session;
+            _fileSystemServices = fileSystemServices;
+            _storageName = name;
+            _dataPath = CreateStoragePath();
+
+            RollbackActions = new List<IRollbackAction>();
+        }
+
+        public RollbackDataStore(ISession session, string name)
+            : this(
+                session,
+                name,
+                new FileSystemServices()
+            )
+        {
+        }
+
+        /// <summary>
+        /// Create, configure, and return the path to be used to store this rollback data
+        /// </summary>
+        private string CreateStoragePath()
+        {
+            var parent = StorageRootPath();
+            var path = Path.Combine(parent, $"{_storageName}.json");
+
+            CreateDirectory(parent);
+
+            return path;
+        }
+
+        /// <summary>
+        /// Create directory @path and secure it so only SYSTEM and Administrators have access
+        /// </summary>
+        /// <param name="path"></param>
+        private void CreateDirectory(string path)
+        {
+            // Create DACL for only SYSTEM and Administrators, disable inheritance
+            FileSystemSecurity security = new DirectorySecurity();
+            security.SetAccessRuleProtection(true, false);
+            security.AddAccessRule(new FileSystemAccessRule(
+                new SecurityIdentifier(WellKnownSidType.LocalSystemSid, null),
+                FileSystemRights.FullControl,
+                InheritanceFlags.ObjectInherit | InheritanceFlags.ContainerInherit,
+                PropagationFlags.None,
+                AccessControlType.Allow));
+            security.AddAccessRule(new FileSystemAccessRule(
+                new SecurityIdentifier(WellKnownSidType.BuiltinAdministratorsSid, null),
+                FileSystemRights.FullControl,
+                InheritanceFlags.ObjectInherit | InheritanceFlags.ContainerInherit,
+                PropagationFlags.None,
+                AccessControlType.Allow));
+
+            // Create the directory with the DACL
+            Directory.CreateDirectory(path, (DirectorySecurity)security);
+        }
+
+        /// <summary>
+        /// Return settings to be used for serializing/deserializing the rollback data
+        /// To avoid deserialization vulnerabilities, use a SerializationBinder to restrict the
+        /// deserializable types.
+        /// </summary>
+        private JsonSerializerSettings GetSerializerSettings() => new()
+        {
+            TypeNameHandling = TypeNameHandling.Objects,
+            SerializationBinder = new KnownTypesBinder
+            {
+                KnownTypes = new List<Type>
+                {
+                    typeof(FilePermissionRollbackInfo)
```

Comment:
can't get the types from the values in `RollbackActions` list because that list will be empty on deserialize during rollback but I can get the types taht implement `IRollbackAction` from the assmbly.